mempool/src/pool.rs

```diff
@@ -0,0 +1,1098 @@
+        */
+        let res = minimum_fee_rate.compute_fee(tx.encoded_size());
+        log::debug!("minimum_mempool_fee for tx: {:?}", res);
+        res
+    }
+
+    fn validate_transaction(&self, tx: &Transaction) -> Result<Conflicts, TxValidationError> {
+        // This validation function is based on Bitcoin Core's MemPoolAccept::PreChecks.
+        // However, as of this stage it does not cover everything covered in Bitcoin Core
+        //
+        // Currently, the items we want covered which are NOT yet covered are:
+        //
+        // - Checking if a transaction is "standard" (see `IsStandardTx`, `AreInputsStandard` in Bitcoin Core). We have yet to decide on Mintlayer's
+        // definition of "standard"
+        //
+        // - Time locks:  Briefly, the corresponding functions in Bitcoin Core are `CheckFinalTx` and
+        // `CheckSequenceLocks`. See mempool/src/time_lock_notes.txt for more details on our
+        // brainstorming on this topic thus far.
+        //
+        // - Bitcoin Core does not relay transactions smaller than 82 bytes (see
+        // MIN_STANDARD_TX_NONWITNESS_SIZE in Bitcoin Core's policy.h)
+        //
+        // - Checking that coinbase inputs have matured
+        //
+        // - Checking the signature operations cost (Bitcoin Core: `GetTransactionSigOpCost`)
+        //
+        // - We have yet to understand and implement this comment pertaining to chain limits and
+        // calculation of in-mempool ancestors:
+        // https://github.com/bitcoin/bitcoin/blob/7c08d81e119570792648fe95bbacddbb1d5f9ae2/src/validation.cpp#L829
+        //
+        // - Bitcoin Core's `EntriesAndTxidsDisjoint` check
+        //
+        // - Bitcoin Core's `PolicyScriptChecks`
+        //
+        // - Bitcoin Core's `ConsensusScriptChecks`
+
+        // Deviations from Bitcoin Core
+        //
+        // - In our FeeRate calculations, we use the `encoded_size`of a transaction. In contrast,
+        // see Bitcoin Core's `CTxMemPoolEntry::GetTxSize()`, which takes into account sigops cost
+        // and witness data. This deviation is not intentional, but rather the result of wanting to
+        // get a basic implementation working. TODO: weigh what notion of size/weight we wish to
+        // use and whether to follow Bitcoin Core in this regard
+        //
+        // We don't have a notion of MAX_MONEY like Bitcoin Core does, so we also don't have a
+        // `MoneyRange` check like the one found in Bitcoin Core's `CheckTransaction`
+
+        if tx.inputs().is_empty() {
+            return Err(TxValidationError::NoInputs);
+        }
+
+        if tx.outputs().is_empty() {
+            return Err(TxValidationError::NoOutputs);
+        }
+
+        let outpoints = tx.inputs().iter().map(|input| input.outpoint()).cloned();
+
+        if has_duplicate_entry(outpoints) {
+            return Err(TxValidationError::DuplicateInputs);
+        }
+
+        if tx.encoded_size() > MAX_BLOCK_SIZE_BYTES {
```

Comment:
Is this check correct? I believe the block size includes the header which is not part of the transaction, right? I also found it a little confusing that it compared the transaction size to maximum block size instead of maximum transaction size which could be set to be equal to `MAX_BLOCK_SIZE_BYTES`.

Reply:
@TheQuantumPhysicist What do you think?